docs/Practices/hub/hub.md

```diff
@@ -0,0 +1,118 @@
+
+## Conventions
+Sometimes guidance that provides more specificity than a guideline needs to exist. 
+
+
+More often than not this is for enabling better implication. For example a naming convention allows engineers to draw better and more accurate conclusions without needing to review documentation.
+
+Additionally guidelines might be library, framework, or project specific. Defining rules around how to interact with a library for example might be pragmatic if that library has capabilities that by some usages might conflict with existing guidelines.
+
+For the purpose of this document, conventions should be seen as more specific guidelines or extensions to the guidelines.
+
+Naming:
+- The entrypoint function for a command or query should start with `command` or `query`.
+- The entrypoint function for a command or query should be in a file called `command.ts` or `query.ts` OR in a region called `Command` or `Query`.
+- Domain code should either be in files named appropriate to the model/process/rule/etc. within a `domain` directory, or be in a `domain.ts` file, or be in a region called `Domain`.
+- Legacy request handlers that are related to a bounded context and are expected to be maintained/iterated should be put in a `legacy-handlers` directory within the appropriate bounded context.
+- Integration code for 3rd party services should be in a `Data` directory.
+- Repository code (code that interacts with a data store), should be in a file called `repository.ts`, OR in a region called `Repository`.
+- (Request or Event) Handlers exposing a command should be in a file called `api.ts`, a region called `API`, or in files called `request.ts` or `event.ts` in a directory called `handlers`.
+- Integration code for bounded contexts in a different project should be in a `dependent-bcs` directory with sub directories matching tha appropriate bounded context. *As of Jan 24 2023 this is not being utilized. Defining this early to reduce decision overhead when this occurs.*
+- CRUD type functions that create new records in a single tab, fetch records by an unique id with no other filters from a single table, or mutate existing records without conditions from a single table, should reside in a directory called `crud-fns` within the shared `repository` directory.
+
+*[Hub Directory Structure Example](examples.md#hub-directory-structure)*
+
+Tests:
+- Test files should sit close to the code they test. *Test files that test code in disparate files should be broken up into multiple test files itself.*
+
+SQL:
+- Avoid using the [massive scripts directory](https://massivejs.org/docs/functions-and-scripts). Prefer instead to define a query as close to it's consumers as possible. *Defining queries in the scripts directory opens up the possibility of unintended coupling by non cohesive modules. There are solutions to syntax highlighting that don't require a .sql file.*
```

Comment:
This [VSCode plugin](https://marketplace.visualstudio.com/items?itemName=frigus02.vscode-sql-tagged-template-literals) is what I found. What else is there?

Reply:
I'll be honest. It's been a while since I've done this.

I used to be able to go to the bottom right of VSCode and change the language rules to be SQL and it would highlight within a string. I just tried this though and it didn't work. Unsure why.

My current process is just taking the SQL and pasting it in a sql client (I use postico).

**Thoughts on this convention**
- While annoying, there are workarounds to the syntax highlighting that a sql file provides.
- The reason we can't put a SQL file right next to each of it's usages - so like right in each slice that needs it - is because massivejs only support a single scripts directory. You can change the location, but it's still a single directory. *I searched for 5 min and couldn't find a way to have multiple scripts dirs. Maybe a more thorough search might turn up something?*
- While it may be pragmatic now and then to bend a convention/guideline/work against a principle because a tool works a certain way, I believe it's best to be extremely selective when you do this to avoid slippery slopes. I think a good rule of thumb is whether it's blocking to be able to get something done. In our case we can still utilize custom queries with massivejs. There's also workaround to get syntax highlighting or a process workaround by developing the query in a sql client. I think efforts in regaining the ergonomics of being able to utilize a SQL file are best done by finding a better tool than massivejs that can utilize SQL files while being able to have them in more than one spot.

Reply:
I tested out the VS code plugin I mentioned and didn't like the way it worked. I'm in favor of a Typed ORM if we can implement it.